dhall/src/Dhall/Set.hs

```diff
@@ -35,13 +35,23 @@ import qualified Data.Set
 import qualified Data.Sequence
 import qualified Data.Foldable
 
+-- Invariant: In @Set set seq@, @toAscList set == sort (toList seq)@.
 data Set a = Set (Data.Set.Set a) (Seq a)
-    deriving (Eq, Generic, Ord, Show, Data, NFData)
+    deriving (Generic, Show, Data, NFData)
+
+instance Eq a => Eq (Set a) where
+    (Set _ x) == (Set _ y) = x == y
```

Comment:
Shouldn't this be done on the first field, not the second? I see in `append` for example, we just snoc onto the list. If I'm reading it correctly, this code means `append x (append y mempty) /= append y (append x mempty)`. Am I missing something?

Reply:
Most of the API is order-sensitive. So the `Eq` instance should also be order-sensitive, right? The [docs for `Eq`](http://hackage.haskell.org/package/base-4.12.0.0/docs/Data-Eq.html#t:Eq) call this "substitutivity".